### DISCUSSION TITLE
Mini-batching in graph level prediction

### DISCUSSION THREAD

**Discussion by ahmedramly** [pyg-team/pytorch_geometric#4778] [Q&A] [answered]
Hi, 
I have a multi-graph dataset, each graph has the same number of nodes and variant number of edges. for each graph,  ```x``` has the size ```num_nodes, feature_dim```.  I am performing graph level prediction, but the problem is that after the GNN layers in the model I don't want to apply pooling layers as it is done ordinarily in this kind of application. instead, I want to have access to ```x``` in its original size after embedding ```num_nodes, embedding_size```, in normal PyTorch mini batching it is easy but I don't know how to do it in PyG mini batching scheme. I  get the output of the model with ```num_nodes_in_batch, embedding_size``` but I need it ```batch, num_nodes, embedding_size```.  I tried to use ```DenseDataLoader``` but the model didn't accept it and also I tried to use ```ListDataLoader``` but I coulfn't figure out how to use it inside the training loop in the correct way. So I tried the approach mentioned in the advanced mini-batching [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html#batching-along-new-dimensions) as follows:
```
class GraphData(Data):
  def __cat_dim__(self, key, value, *args, **kwargs):
    if key == 'feats':
      return None
    else:
      return super().__cat_dim__(key, value, *args, **kwargs)
```
and constructed each graph as follows:
```data = GraphData(feats=x, edge_index=edge_index, edge_attr=edge_attr, y=y, num_nodes=x.shape[0])```

the batch looks fine 
```GraphDataBatch(edge_index=[2, 26432], edge_attr=[26432, 40], y=[64], feats=[64, 64, 641], num_nodes=4096, batch=[4096], ptr=[65])```
 

But when I fed the data into the trainer it raised this error:
```RuntimeError: The size of tensor a (4096) must match the size of tensor b (64) at non-singleton dimension 1```

I am using  a model like this one for a classification task and I need to perform further computations on each sample ```num_nodes, embedding_size``` NOT just ```average pooling``` or ```max pooling```

```
class GNNTransformer(pl.LightningModule):

  def __init__(self,
               feature_size,
               embedding_size,
               **kwargs):
    
    super(GNNTransformer, self).__init__()

    self.save_hyperparameters()
    self.conv_1 = TransformerConv(feature_size,
                                 embedding_size,
                                 ...)
    self.transf_1 = nn.Linear(n_heads*embedding_size, embedding_size)
  
    self.other_layers = ...   #input should be 'batch, n_nodes, embedding_size'
    self.fc = nn.Linear(some_dim, num_classes)

  def forward(self, x, edge_index, edge_attr, batch_idx):
    # First Layer
    x = self.conv_1(x, edge_index, edge_attr)
    x = torch.relu(self.transf_1(x))

    x = self.other_layers(x) #?????
    x = self.fc(x)
    x = F.log_softmax(x, dim=1)
    return x
```
how could I solve this issue?

**Comment by rusty1s** [comment, marked as answer]
Please have a look at [`to_dense_batch`](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.to_dense_batch).

**Reply by ahmedramly** [reply]
Thanks a lot, I was not aware of this function.
